_posts/2020-10-01-torch-network-from-scratch/torch_network_from_scratch.Rmd

```diff
@@ -0,0 +1,1076 @@
+     0.1386 -0.6107 -1.2527 -1.3256 -0.1009
+    [ CPUFloatType{3,5} ]
+
+If it were of size `5`, a virtual leading dimension would be added, and then, the same broadcasting would take place as in the previous case.
+
+```{r}
+t1 <- torch_randn(c(3,5))
+t2 <- torch_randn(c(5))
+
+t1$add(t2)
+```
+
+    torch_tensor 
+    -1.4123  2.1392 -0.9891  1.1636 -1.4960
+     0.8147  1.0368 -2.6144  0.6075 -2.0776
+    -2.3502  1.4165  0.4651 -0.8816 -1.0685
+    [ CPUFloatType{3,5} ]
+
+Here is a more complex example. Broadcasting how happens both in `t1` and in `t2`:
+
+```{r}
+t1 <- torch_randn(c(1,5))
+t2 <- torch_randn(c(3,1))
+
+t1$add(t2)
+```
+
+    torch_tensor 
+     1.2274  1.1880  0.8531  1.8511 -0.0627
+     0.2639  0.2246 -0.1103  0.8877 -1.0262
+    -1.5951 -1.6344 -1.9693 -0.9713 -2.8852
+    [ CPUFloatType{3,5} ]
+
+As a nice concluding example, through broadcasting, an outer product can be computed like so:
+
+```{r}
+t1 <- torch_tensor(c(0, 10, 20, 30))
+
+t2 <- torch_tensor(c(1, 2, 3))
+
+t1$view(c(4,1)) * t2
+```
+
+    torch_tensor 
+      0   0   0
+     10  20  30
+     20  40  60
+     30  60  90
+    [ CPUFloatType{4,3} ]
+
+And now, we really get to implementing that neural network!
+
+## A simple neural network using `torch` tensors
+
+Our task, which we'll approach in a low-level way today but considerably simplify in upcoming installments, consists of regressing a single target datum based on three input variables.
+
+We directly use `torch` to simulate some data.
+
+#### Toy data
+
+```{r}
+library(torch)
+
+# input dimensionality (number of input features)
+d_in <- 3
+# output dimensionality (number of predicted features)
+d_out <- 1
+# number of observations in training set
+n <- 100
+
+
+# create random data
+# input
+x <- torch_randn(n, d_in)
+# target
+y <- x[, 1, drop = FALSE] * 0.2 -
+  x[, 2, drop = FALSE] * 1.3 -
+  x[, 3, drop = FALSE] * 0.5 +
+  torch_randn(n, 1)
+```
+
+Next, we need to initialize the network's weights. We'll have one hidden layer, with `32` units. The output layer's size, being determined by the task, is equal to `1`.
+
+#### Initialize weights
+
+```{r}
+# dimensionality of hidden layer
+d_hidden <- 32
+
+# weights connecting input to hidden layer
+w1 <- torch_randn(d_in, d_hidden)
+# weights connecting hidden to output layer
+w2 <- torch_randn(d_hidden, d_out)
+
+# hidden layer bias
+b1 <- torch_zeros(1, d_hidden)
+# output layer bias
+b2 <- torch_zeros(1, d_out)
+```
+
+Now for the training loop proper. The training loop really here *is* the network.
+
+#### Training loop
+
+In each iteration ("epoch"), the training loop does four things:
+
+-   run through the network, computing predictions (*forward pass)*
+
+-   compare those predictions to the ground truth and quantify the loss
+
+-   run backwards through the network, computing the gradients that indicate how the weights should be changed
+
+-   actually update the weights, making use of the requested learning rate
```

Comment:
```suggestion
-   updates the weights, making use of the requested learning rate.
```

Reply:
Changed all these to present tense.

Reply:
hm...

> In each iteration ("epoch"), the training loop does four things:

my feeling was that each of those sentences was saying sth like
- does run
- does update
...

but probably this again is some non-native speaker's thing